mlflow/server/handlers.py

```diff
@@ -1173,37 +1184,50 @@ def get_metric_history_bulk_interval_handler():
 
     store = _get_tracking_store()
 
-    def _get_max_step_for_metric(run_id, metric_key):
-        if hasattr(store, "get_max_step_for_metric"):
-            return store.get_max_step_for_metric(run_id=run_id, metric_key=metric_key)
-        steps = [m.step for m in store.get_metric_history(run_id, metric_key)]
-        return max(steps) if steps else 0
-
     def _get_sampled_steps(run_ids, metric_key, max_results):
         # cannot fetch from request_message as the default value is 0
         start_step = args.get("start_step")
         end_step = args.get("end_step")
-        if start_step is None and end_step is None:
-            max_steps_per_run = [_get_max_step_for_metric(run_id, metric_key) for run_id in run_ids]
-            max_steps = max(max_steps_per_run)
-            # Get sampled steps and append the max step for each run.
-            sampled_steps = _get_sampled_steps_from_steps(0, max_steps, max_results)
-            return list(set(sampled_steps).union(max_steps_per_run))
-        elif start_step is not None and end_step is not None:
+
+        # perform validation before any data fetching occurs
+        if start_step is not None and end_step is not None:
             start_step = int(start_step)
             end_step = int(end_step)
             if start_step > end_step:
                 raise MlflowException.invalid_parameter_value(
                     "end_step must be greater than start_step. "
                     f"Found start_step={start_step} and end_step={end_step}."
                 )
-            sampled_steps = _get_sampled_steps_from_steps(start_step, end_step, max_results)
-            return list(set(sampled_steps).union([end_step]))
-        else:
+        elif start_step is not None or end_step is not None:
             raise MlflowException.invalid_parameter_value(
                 "If either start step or end step are specified, both must be specified."
             )
 
+        # get a list of all steps for all runs. this is necessary
+        # because we can't assume that every step was logged, so
+        # sampling needs to be done on the steps that actually exist
+        all_runs = [
+            [m.step for m in store.get_metric_history(run_id, metric_key)] for run_id in run_ids
+        ]
+
+        # save mins and maxes to be added back later
+        all_mins_and_maxes = {step for run in all_runs if run for step in [min(run), max(run)]}
+        all_steps = sorted({step for sublist in all_runs for step in sublist})
```

Comment:
Yup, this is logic is aligned with the databricks backend. It might miss some cases where a given metric is logged at different intervals between runs, but that case should be unlikely.